app/upgrades/v17/upgrades.go

```diff
@@ -133,3 +145,35 @@ func CreateUpgradeHandler(
 		return migrations, nil
 	}
 }
+
+func FlipTwapSpotPriceRecords(ctx sdk.Context, poolIds []uint64, keepers *keepers.AppKeepers) error {
+	for _, poolId := range poolIds {
+		// check that this is a cl pool
+		_, err := keepers.ConcentratedLiquidityKeeper.GetConcentratedPoolById(ctx, poolId)
+		if err != nil {
+			return err
+		}
+
+		// check that the twap record exists
+		clPoolTwapRecords, err := keepers.TwapKeeper.GetAllMostRecentRecordsForPool(ctx, poolId)
+		if err != nil {
+			return err
+		}
+
+		for _, twapRecord := range clPoolTwapRecords {
+			twapRecord.LastErrorTime = time.Time{}
+			oldAsset0Denom := twapRecord.Asset0Denom
+			oldAsset1Denom := twapRecord.Asset1Denom
+			oldSpotPrice0 := twapRecord.P0LastSpotPrice
+			oldSpotPrice1 := twapRecord.P1LastSpotPrice
+
+			twapRecord.Asset0Denom = oldAsset1Denom
+			twapRecord.Asset1Denom = oldAsset0Denom
+			twapRecord.P0LastSpotPrice = oldSpotPrice1
+			twapRecord.P1LastSpotPrice = oldSpotPrice0
+
+			keepers.TwapKeeper.StoreNewRecord(ctx, oldAsset0Denom, oldAsset1Denom, twapRecord)
+		}
+	}
+	return nil
```

Comment:
Similar to most recent records, we should also be getting the two kinds of historical records that are written at separate store indexes:

https://github.com/osmosis-labs/osmosis/blob/74d8aa886be5d2ae1757bea8d76ddb94fcecb53e/x/twap/store.go#L155-L163

Reply:
we updated `StoreNewRecord`'s TWAPRecord which updates values for historicalTimeindexed and historicalPoolIndex no? We donot change the keys just update the value

Reply:
It updates them by adding 2 new records. One for each of the historical indexes.

However, the old historical records still remain switched

Reply:
okay so here's what i did; 
1. Update `HistoricalTimeIndexedTWAPs` and  Delete old `HistoricalTimeIndexedTWAPs`
2. Update `HistoricalPoolIndexedTWAPs` and Delete old `HistoricalPoolIndexedTWAPs`
3. Update `MostRecentTWAP` Delete old `MostRecentTWAP` 
  - here interestingly what also happens is we call `k.StoreHistoricalTWAP(ctx, twap)` which reset the **corrected** `HistoricalTimeIndexedTWAPs` `HistoricalPoolIndexedTWAPs` and makes it incorrect. 
  
  one way to solve this would be to just delete `HistoricalTimeIndexedTWAPs`  and `HistoricalPoolIndexedTWAPs`

Reply:
actually i may have found a solution. I skipped (2) and just did 1 and 3 and the results are accurate 

old issue with HistoricalRecords (I printed both GetAllHistoricalPoolIndexed & TimeIndexedTWAPs) : 
```
PreUpgrade
 [{1 ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4 uosmo 0 2023-08-03 21:38:45.722480581 +0000 UTC 10 0.1 100 3 ...} {1 ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4 uosmo 1 2023-08-04 21:38:45.703693435 +0000 UTC 0 0 0 0 0 ...}]

PostUpgrade 

[{1 ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4 uosmo 0 2023-08-03 21:38:45.722480581 +0000 UTC 10 0.1 100 3 ...} {1 uosmo ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4 0 2023-08-03 21:38:45.722480581 +0000 UTC 0.1 10 100 3 ... {1 ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4 uosmo 1 2023-08-04 21:38:45.703693435 +0000 UTC 0 0 0 0 0 ...}]

*** WITH ABOVE FIX ***

PRE UPGRADE: 
[{1 ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4 uosmo 0 2023-08-03 21:42:14.531768482 +0000 UTC 10 0.1 100 3 0 ...} {1 ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4 uosmo 1 2023-08-04 21:42:14.514891373 +0000 UTC 0 0 0 0 0 ...]

POST UPGRADE:  
[{1 uosmo ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4 0 2023-08-03 21:42:14.531768482 +0000 UTC 0.1 10 100 3 ... {1 uosmo ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4 1 2023-08-04 21:42:14.514891373 +0000 UTC 0 ...}]
```
which is exactly what we want. lmk what you think

Reply:
Makes sense. However, please see my other comment that points out all twaps are being changed now. As a result, the original 1), 2), 3) approach is probably the way with one modification.

The modification is to introduce a `StoreMostRecentRecord` that we can use to work around the issue of calling `StoreHistoricalTWAP`.

Please make sure that `StoreRecord` uses this new helper as well to reduce duplication

Reply:
discussed offline